### PR TITLE
add  cl_khr_expect_assume to test_compiler_defines_for_extensions

### PR DESCRIPTION
**Pull request by norablackcat**
https://github.com/KhronosGroup/OpenCL-CTS/issues/1424#issuecomment-1554680436

### CODE REVIEW

File: test_conformance/compiler/test_compiler_defines_for_extensions.cpp [hunk 1]
```diff
@@ -63,6 +63,7 @@ const char *known_extensions[] = {
     "cl_khr_dx9_media_sharing",
     "cl_khr_egl_event",
     "cl_khr_egl_image",
+    "cl_khr_expect_assume",
```
Comment (svenvh) [inline test_conformance/compiler/test_compiler_defines_for_extensions.cpp:66]: As this isn't an API-only extension, I believe this should go right before `cl_khr_subgroup_rotate`. See also the comments on lines 23 and 56.

Please also update `first_API_extension` accordingly.

**Review by svenvh — CHANGES_REQUESTED** [review]

**Review by svenvh — DISMISSED** [review]
LGTM, thanks!

**Review by lakshmih — APPROVED** [review]

**Review by bashbaug — APPROVED** [review]
Thank you!

**Review by bashbaug — COMMENTED** [review]
Actually, I'm sorry, I did have one question/comment: because this extension does not extend OpenCL C, is it really correct to consider it to be a non-API-only extension?

As-written, this test will fail if an implementation reports support for `cl_khr_expect_assume` but the proper support is NOT "defined in the OpenLC C kernel environment", which does not quite seem correct.

**Comment by norablackcat** [comment]
I actually had it as an api only extension at first because the extension spec doesn't define any feature macros for it. It does seem a bit silly to me to define cl_khr_expect_assume macro that doesn't do anything because spirv does the work. at least in the rusticl implementation.

**Comment by karolherbst** [comment]
There are a handful of SPIR-V only extensions now and I suspect we might want to figure how to treat those.

"API only" here basically means, that the online compiler of the OpenCL runtime doesn't expose any feature macro and I think that term is therefore misleading.

So if an extension doesn't define anything on the OpenCL C level, we should just treat them as "API only". Some extensions might define a feature macro in additions to new OpenCL C features and we'll have to check those explicitly anyway.

**Comment by karolherbst** [comment]
Side note: I don't think we should allow SPIR-V only extensions under `khr`, because they just fragment how compilers actually implement those extensions in their OpenCL C frontend and we might end up with implementations doing it differently. We should rather just define this alongside the SPIR-V side so it's all consistent across various OpenCL C to SPIR-V compilers, which in almost every case will be clang anyway)

**Comment by bashbaug** [comment]
> If I had a say in it I would say it's api only.

Yeah, I think I agree with this too, though I'm starting to put "api only" in quotes.

I think what really happened here is that we used to only have two categories of extensions: "API-only" extensions, which only affected OpenCL runtime APIs, and "non-API-only" extensions, which affected the (OpenCL C) compiler.  The first category did not need any macros, because they only affected runtime APIs, but the second did.  This is how the two categories of extensions are used in this test.

We've since added a third category of extensions that affect the compiler, but does not affect the OpenCL C compilers, and hence does not need or include a macro.  These aren't really "API-only" extensions, but this categorization works for the purposes of this test.

> Side note: I don't think we should allow SPIR-V only extensions under khr, because they just fragment how compilers actually implement those extensions in their OpenCL C frontend and we might end up with implementations doing it differently. We should rather just define this alongside the SPIR-V side so it's all consistent across various OpenCL C to SPIR-V compilers, which in almost every case will be clang anyway

Can you say a little more about how this might work?  I'd love to have a way to streamline SPIR-V-only extensions but I haven't thought of a clever way to do this just yet.

Note that some of these SPIR-V-only extensions may not have any representation in OpenCL C (`cl_khr_spirv_linkonce_odr` is a good example), but they're still useful so kernels written in other source languages may executed via an OpenCL runtime.

We should probably move this conversation to a separate issue so it doesn't disappear when this PR gets merged.

**Comment by karolherbst** [comment]
@bashbaug https://github.com/KhronosGroup/OpenCL-Docs/issues/941

**Comment by karolherbst** [comment]
btw, we already treat `cl_khr_spirv_no_integer_wrap_decoration` as API only, which is also spirv only in its current form.

**Review by bashbaug — CHANGES_REQUESTED** [review]
Just so this doesn't get merged by accident: we should be considering this an "API-only" extension because it does not formally affect OpenCL C, even though it is really primarily a SPIR-V extension.  See discussion above.

**Review by svenvh — APPROVED** [review]
@bashbaug the latest version treats this as an API-only extension again.  I suppose we can merge it now or do you have any further concerns?

**Review by bashbaug — APPROVED** [review]

**Comment by bashbaug** [comment]
All good from my end, thanks for the nudge!

**Comment by bashbaug** [comment]
This has two approvals now, plus @lakshmih's prior approval.  Merging.

**Pull request merged**